fms/utils/serialization.py

```diff
@@ -55,6 +59,96 @@ def list_sources(architecture: str):
     return list(__adapters[architecture].keys())
 
 
+def _register_legacy_weight_preprocessor(
```

Comment:
With the repo less than a year old, I don't think we should be creating a registration mechanism for a single older weight format. If we change things again, we'll end up with multiple levels of "legacy" and the meaning will be ambiguous.

I think we could either interpret the old weights as an alternate "source" of weights. We could probably also version them. (e.g. we have fms, hf, meta, ... we could have fms.v1 and fms.v2 and treat fms/None as an alias to fms.(whatever is latest)).

Reply:
A user may not know which weight version they have stored, all this function is doing is registering a preprocessor for a given model to fix the weights to a new version. 

Example:

In 0.0.3 Query, key, value gets changed to qkv_fused

 - Register a new preprocessor for query, key, and value to modify to qkv_fused

In 0.1.0 qkv_fused gets changed to attn.qkv_fused:

- Register a new preprocessor for qkv_fused to modify to attn.qkv_fused (Note: we are still registering the 0.0.3 preprocessor, none of that changes)

In 0.1.2 attn.qkv_fused format changes such that the weights tensor has been transposed

- Register a new preprocessor for attn.qkv_fused to modify the weights (Note: we are still registering the 0.1.0 change)

Maybe what we need here is a set of preprocessors that will (when in the correct order), convert to the latest versions set of weights.


So in the above example, if for instance, a user has weights from version 0.0.4:

1. the legacy weight conversion from 0.0.3 would be a no-op (as it would not find query, key, and value)
2. The legacy weight conversion from 0.1.0 would return the new attn.qkv_fused
3. The legacy weight conversion from 0.1.2 will take the output of (2) and format it to the latest version

In this case, we will not require a user to know what version their weights have come from, it is the responsibility of the fms repo to make sure that they registered the preprocessors in the correct order such that weights always end up in the latest version. This will incur a very small overhead during weight loading as you may convert the weights more than once, however id be surprised if it is noticeable. I don't believe this will end up being ambiguous as long as the ordering is preserved.

Let me know your thoughts?

Reply:
As I mentioned offline yesterday, this is a similar concept to Django migrations for databases: https://docs.djangoproject.com/en/5.0/topics/migrations/. I used that extensibly in the past and it worked pretty well, so in principle I am not opposed to it. Instead of returning None, given that we already have flow-control using exceptions (with the FusableWeights exception, we might get away with adding another one so that what happens is made clearer.